lib/auth/version.go

```diff
@@ -20,55 +20,101 @@ package auth
+					continue
+				}
+				return nil
+			} else if err != nil {
+				return trace.Wrap(err)
+			}
+
+			authInfo, err = authinfo.NewAuthInfo(&authinfov1.AuthInfoSpec{TeleportVersion: lastKnownVersion.String()})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if _, err := backendStorage.CreateAuthInfo(ctx, authInfo); err != nil {
+				err = trace.Wrap(err)
+				slog.WarnContext(ctx, "Failed to create AuthInfo resource",
+					"error", err)
+				continue
+			}
+		} else if err != nil {
 			return trace.Wrap(err)
 		}
-		return nil
-	} else if err != nil {
-		return trace.Wrap(err)
-	}
 
-	if currentVersion.Major-lastKnownVersion.Major > 1 {
-		return trace.BadParameter("Unsupported upgrade path detected: from %v to %v. "+
-			"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
-			"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
-			"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
-			lastKnownVersion, currentVersion.String(), lastKnownVersion.Major+1)
-	}
-	if lastKnownVersion.Major-currentVersion.Major > 1 {
-		return trace.BadParameter("Unsupported downgrade path detected: from %v to %v. "+
-			"Teleport doesn't support major version downgrade.\n Please downgrade "+
-			"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
-			"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
-			lastKnownVersion, currentVersion.String(), lastKnownVersion.Major-1)
-	}
-	if err := storage.WriteTeleportVersion(ctx, currentVersion); err != nil {
-		return trace.Wrap(err)
+		lastKnownVersion, err := semver.NewVersion(authInfo.GetSpec().GetTeleportVersion())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if currentVersion.Major-lastKnownVersion.Major > 1 {
+			return trace.BadParameter("Unsupported upgrade path detected: from %v to %v. "+
+				"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
+				"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
+				"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
+				lastKnownVersion, currentVersion.String(), lastKnownVersion.Major+1)
+		}
+		if lastKnownVersion.Major-currentVersion.Major > 1 {
+			return trace.BadParameter("Unsupported downgrade path detected: from %v to %v. "+
+				"Teleport doesn't support major version downgrade.\n Please downgrade "+
+				"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
+				"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
+				lastKnownVersion, currentVersion.String(), lastKnownVersion.Major-1)
+		}
+
+		authInfo.GetSpec().TeleportVersion = currentVersion.String()
+		if _, err := backendStorage.UpdateAuthInfo(ctx, authInfo); err != nil {
+			err = trace.Wrap(err)
+			slog.WarnContext(ctx, "Failed to update AuthInfo resource",
+				"error", err)
+			continue
+		}
```

Comment:
We only need to write the version if it's different (higher?), this write can be skipped if the version is the same.